modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -111,21 +108,49 @@ def __init__(
 
         self._final_checkpoint_path.mkdir()  # exist_ok == False, this directory should not exist before
 
-        self._status_query_queue = status_query_queue
-        self._status_response_queue = status_response_queue
+        self._status_query_queue_training = status_query_queue_training
+        self._status_response_queue_training = status_response_queue_training
+
+        self._status_query_queue_downsampling = status_query_queue_downsampling
+        self._status_response_queue_downsampling = status_response_queue_downsampling
 
         self._num_samples = 0
 
         self._metadata_collector = MetadataCollector(training_info.pipeline_id, training_info.trigger_id)
 
         self.selector_stub = self.connect_to_selector(training_info.selector_address)
-        self._downsampling_enabled, strategy_name, params_from_selector = self.get_selection_strategy()
-        self._weighted_opt = False
+        self.selector_address = training_info.selector_address
+        downsampling_enabled, strategy_name, downsampler_config = self.get_selection_strategy()
 
-        if self._downsampling_enabled:
+        if downsampling_enabled:
             self._criterion_nored = criterion_func(**training_info.criterion_dict, reduction="none")
-            self._downsampler = self.instantiate_downsampler(strategy_name, params_from_selector, self._criterion_nored)
-            self._weighted_opt = True
+            self._downsampler = self.instantiate_downsampler(strategy_name, downsampler_config, self._criterion_nored)
+
+            assert "sample_then_batch" in downsampler_config
+            if downsampler_config["sample_then_batch"]:
+                self._downsampling_mode = DownsamplingMode.SAMPLE_THEN_BATCH
+                assert "downsampling_period" in downsampler_config
+                self._downsampling_period = downsampler_config["downsampling_period"]
+                self.offline_dataset_path = training_info.offline_dataset_path
+            else:
+                self._downsampling_mode = DownsamplingMode.BATCH_THEN_SAMPLE
+        else:
+            self._downsampling_mode = DownsamplingMode.DISABLED
```

Comment:
just for readability: please move this into self._setup_downsampling or something, so init becomes a bit smaller

Reply:
I only left this in the `__init__`.
```
downsampling_enabled, strategy_name, downsampler_config = self.get_selection_strategy()
  if downsampling_enabled:
      self._setup_downsampling(criterion_func, downsampler_config, strategy_name, training_info)
  else:
      self._downsampling_mode = DownsamplingMode.DISABLED
```

Reply:
thanks!